File: src/main/java/com/smashingmods/alchemistry/compat/jei/category/AtomizerRecipeCategory.java
package com.smashingmods.alchemistry.compat.jei.category;

import com.smashingmods.alchemistry.Alchemistry;
import com.smashingmods.alchemistry.Registration;
import com.smashingmods.alchemistry.blocks.atomizer.AtomizerRecipe;
import com.smashingmods.alchemistry.compat.jei.JEIIntegration;
import mezz.jei.api.constants.VanillaTypes;
import mezz.jei.api.gui.IRecipeLayout;
import mezz.jei.api.gui.drawable.IDrawable;
import mezz.jei.api.gui.ingredient.IGuiFluidStackGroup;
import mezz.jei.api.gui.ingredient.IGuiItemStackGroup;
import mezz.jei.api.helpers.IGuiHelper;
import mezz.jei.api.ingredients.IIngredients;
import mezz.jei.api.recipe.category.IRecipeCategory;
import net.minecraft.client.resources.language.I18n;
import net.minecraft.network.chat.Component;
import net.minecraft.network.chat.TextComponent;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.item.ItemStack;
import net.minecraftforge.fluids.FluidStack;

public class AtomizerRecipeCategory implements IRecipeCategory<AtomizerRecipe> {
    private IGuiHelper guiHelper;
    private final static int u = 40;
    private final static int v = 11;

    public AtomizerRecipeCategory(IGuiHelper guiHelper) {
        this.guiHelper = guiHelper;
    }

    @Override
    public ResourceLocation getUid() {
        return new ResourceLocation(Alchemistry.MODID, JEIIntegration.ATOMIZER_CATEGORY);
    }

    @Override
    public Class<? extends AtomizerRecipe> getRecipeClass() {
        return AtomizerRecipe.class;
    }

    @Override
    public Component getTitle() {
        return new TextComponent(I18n.get("alchemistry.jei.atomizer"));
    }

    @Override
    public IDrawable getBackground() {
        return guiHelper.createDrawable(new ResourceLocation(Alchemistry.MODID, "textures/gui/atomizer_gui.png"),
                u, v, 100, 125);
    }

    @Override
    public IDrawable getIcon() {
        return guiHelper.createDrawableIngredient(new ItemStack(Registration.ATOMIZER_BLOCK.get()));
    }

    @Override
    public void setIngredients(AtomizerRecipe recipe, IIngredients ingredients) {
        ingredients.setInput(VanillaTypes.FLUID, recipe.input);
        ingredients.setOutput(VanillaTypes.ITEM, recipe.output);
    }

    @Override
    public void setRecipe(IRecipeLayout recipeLayout, AtomizerRecipe recipe, IIngredients ingredients) {
        IGuiItemStackGroup guiItemStacks = recipeLayout.getItemStacks();
        IGuiFluidStackGroup guiFluidStacks = recipeLayout.getFluidStacks();


        int x = 121 - u;
        int y = 51 - v;
        guiItemStacks.init(0, false, x, y);
        guiItemStacks.set(0, ingredients.getOutputs(VanillaTypes.ITEM).get(0));

        x = 48 - u;
        y = 69 - u;
        FluidStack inputStack = ingredients.getInputs(VanillaTypes.FLUID).get(0).get(0);
        guiFluidStacks.init(1, true, x, y, 16, 60, inputStack.getAmount(), false, null);
        guiFluidStacks.set(1, inputStack);
    }
}

File: src/main/kotlin/al132/alchemistry/compat/jei/atomizer/AtomizerRecipeCategory.kt
package al132.alchemistry.compat.jei.atomizer

import al132.alchemistry.client.GuiAtomizer
import al132.alchemistry.compat.jei.AlchemistryRecipeCategory
import al132.alchemistry.compat.jei.AlchemistryRecipeUID
import al132.alchemistry.compat.jei.Translator
import mezz.jei.api.IGuiHelper
import mezz.jei.api.gui.IRecipeLayout
import mezz.jei.api.ingredients.IIngredients
import mezz.jei.api.ingredients.VanillaTypes

class AtomizerRecipeCategory(guiHelper: IGuiHelper)
    : AlchemistryRecipeCategory<AtomizerRecipeWrapper>(guiHelper.createDrawable(guiTexture, u, v, 100, 125),
        "jei.atomizer.name") {

    override fun getTitle(): String = Translator.translateToLocal("jei.atomizer.name")

    override fun getUid(): String = AlchemistryRecipeUID.ATOMIZER

    override fun setRecipe(recipeLayout: IRecipeLayout, recipeWrapper: AtomizerRecipeWrapper, ingredients: IIngredients) {
        val guiItemStacks = recipeLayout.itemStacks
        val guiFluidStacks = recipeLayout.fluidStacks


        var x = 121 - u
        var y = 51 - v
        guiItemStacks.init(OUTPUT_ONE, false, x, y)
        guiItemStacks.set(OUTPUT_ONE, ingredients.getOutputs(VanillaTypes.ITEM)[0])

        x = 48 - u
        y = 69 - u
        val inputStack = ingredients.getInputs(VanillaTypes.FLUID)[0][0]
        guiFluidStacks.init(FLUID_ONE, true, x, y, 16, 60, inputStack.amount, false, null)
        guiFluidStacks.set(FLUID_ONE, inputStack)
    }

    companion object {

        private val OUTPUT_ONE = 1
        private val FLUID_ONE = 1

        private val u = 40
        private val v = 11

        private val guiTexture = GuiAtomizer.textureLocation
    }
}